clients/src/main/java/org/apache/kafka/common/record/DefaultRecord.java

```diff
@@ -370,6 +372,99 @@ private static DefaultRecord readFrom(ByteBuffer buffer,
         }
     }
 
+    public static DefaultRecord readPartiallyFrom(DataInput input,
+                                                  long baseOffset,
+                                                  long baseTimestamp,
+                                                  int baseSequence,
+                                                  Long logAppendTime) throws IOException {
+        int sizeOfBodyInBytes = ByteUtils.readVarint(input);
+        int totalSizeInBytes = ByteUtils.sizeOfVarint(sizeOfBodyInBytes) + sizeOfBodyInBytes;
+
+        // only use skipBytes when the size is large enough to be beneficial
+        if (sizeOfBodyInBytes > MAX_SKIP_BUFFER_SIZE * 3) {
+            return readPartiallyFrom(input, totalSizeInBytes, sizeOfBodyInBytes, baseOffset, baseTimestamp,
+                baseSequence, logAppendTime);
+        } else {
+            ByteBuffer recordBuffer = ByteBuffer.allocate(sizeOfBodyInBytes);
+            input.readFully(recordBuffer.array(), 0, sizeOfBodyInBytes);
+            return readFrom(recordBuffer, totalSizeInBytes, sizeOfBodyInBytes, baseOffset, baseTimestamp,
+                baseSequence, logAppendTime);
+        }
+    }
+
+    private static PartialDefaultRecord readPartiallyFrom(DataInput input,
```

Comment:
I've been thinking about this function a little bit and the savings we are trying to achieve. From a high level, I think we are trying to minimize the number of reads and the number of allocations. I am considering whether we can pull the `skipBytes` approach up to this level. So first thing we do is allocate a buffer of size `min(sizeInBytes, FIXED_BUFFER_SIZE)`. Then we turn the parsing into a state machine. The logic would look something like this:

```
buffer = allocate()
needMore = true
state = READ_METADATA

while (true) {
 if (needMore)
  input.read(buffer)

 switch (state) {
  case READ_METADATA:
    // read offset and timestamp
    // transition to READ_KEY
  case READ_KEY_SIZE:
    // read key size
    // transition to SKIP_KEY_BYTES
  case SKIP_KEY_BYTES:
    // skip until key is consumed
    // transition to READ_VALUE
  case READ_VALUE_SIZE:
    // read value size
    // transition to SKIP_VALUE_BYTES
  case SKIP_VALUE_BYTES:
    // skip bytes until value is consumed
    // transition to READ_HEADERS
  case READ_NUM_HEADERS:
    // read num headers
    // transition to READ_HEADER
  case READ_HEADER:
    // read one header
    // return when all expected headers are consumed
 } 
}
```
The basic idea is to never have any trivial reads. In fact, this would allow us to reuse the same buffer across multiple calls to `readPartially` which would improve GC even further.